downstairs/src/lib.rs

```diff
@@ -1473,14 +1485,133 @@ impl Downstairs {
         }
     }
 
-    // Downstairs, given a job ID, do the work for that IO.
+    // Given a job ID, do the work for that IO.
+    //
+    // This method calls into the Downstair's region and performs the read /
+    // write / flush action.
     async fn do_work(
-        &self,
+        &mut self,
         upstairs_connection: UpstairsConnection,
         job_id: u64,
     ) -> Result<Option<Message>> {
-        let mut work = self.work_lock(upstairs_connection).await?;
-        work.do_work(self, job_id).await
+        let job = {
+            let mut work = self.work_lock(upstairs_connection).await?;
+            let job = work.get_ready_job(job_id).await;
+
+            // `promote_to_active` can clear out the Work struct for this
+            // UpstairsConnection, but the tasks can still be working on
+            // outdated job IDs. If that happens, `get_ready_job` will return a
+            // None, so bail early here.
+            if job.is_none() {
+                return Ok(None);
+            }
+
+            job.unwrap()
+        };
+
```

Comment:
The old do_work() had a few asserts here:
```
    assert_eq!(job.state, WorkState::InProgress);
    assert_eq!(job_id, job.ds_id);
```
Would those still make sense?

It also had a check to be sure the deps were satisfied.  Same question for that.

Reply:
Ah, nope, nevermind.
I see this now happens in the `get_ready_job()` method.
So, why is this comment even here still?  Because I want you to see how hard I'm working, that's why.